<!DOCTYPE HTML PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd"><html xmlns="http://www.w3.org/1999/xhtml">
<head><title>Amira Online Help</title>
<meta http-equiv="Content-Type" content="text/html;charset=utf-8" />
<link rel="stylesheet" href="../Amira.css" type="text/css" />
<script type="text/javascript" src="../MathJax/MathJax.js?config=default.js"></script>
</head>
<body lang="en">


<p></p><div class="generic_head"><h2><a name="A1"> Molecule</a>
</h2> <ul class="entry_classification"><li>Data Type</li><li>Molecular Option</li></ul></div>
<! HxIndex: "amiraMol" "Data" "Molecule" "a single molecular structure" !>
<p></p><div class="hxdescription">
<h3>Description</h3>
<p>
 An object of data type <i>Molecule</i> contains information about the
 structure of the molecule. Typical information that is stored are
 the types and positions of the molecule's atoms and the bonds
 between the atoms, plus the type of each bond. Furthermore, the data
 object stores information about groups of atoms in a hierarchical
 way. For example a functional group consists of a number of atoms,
 several functional groups may form a residue, and a couple of
 residues may form a secondary structure. This allows quick traversal
 of the molecule's structure.</p>
<p> Molecules can be loaded from a file or generated by several
 <a class="link" href="molglobaltclcommands.html#A1">global tcl commands</a>.
 They can be edited by either using some of the data objects
 tcl commands which are described in the following sections or by using the
 <a class="link" href="../hxmoledit/HxMoleculeEditor.html#A1">Molecule-Editor</a>,
 <a class="link" href="../hxmoledit/HxAttributeEditor.html#A1">Attribute-Editor</a>,
 or <a class="link" href="../hxmoledit/HxDataEditor.html#A1">Data-Editor</a>.</p>
<p> Molecules can be visualized with the following viewing modules:
 <a class="link" href="../hxmolview/HxMolView.html#A1">MoleculeView</a>, <a class="link" href="../hxmolview/HxBondAngle.html#A1">BondAngleView</a>,
 <a class="link" href="../hxmolview/HxSecStructure.html#A1">SecStructureView</a>, or
 <a class="link" href="../hxmolview/HxTubeView.html#A1">TubeView</a>. In addition, there are two modules for
 generating molecular surfaces. The
 <a class="link" href="../hxmolsurface/HxCompMolSurface.html#A1">CompMolSurface</a> module enables you to
 generate the <i>solvent accessible, solvent excluded</i>, and <i>van der Waals surfaces</i> of a molecule. The
 <a class="link" href="../hxmolsurface/HxCompMolInterface.html#A1">CompMolInterface</a> module can be used to
 generate intra- and intermolecular interfaces, such as between
 single atoms or residues, or between two molecules, respectively.</p>
<p> When comparing the structures of several molecules with each other,
 one is faced with the problem of aligning molecules to each other.
 This can be easily done
 
 
 
 
 by connecting the <em>AlignMaster</em> connection port to a second
 molecule which will serve as reference. There are several alignment
 modes. If the molecules have the same number of atoms you can align
 the two molecules to each other by using all atoms, whereby the i'th
 atom of the first molecule corresponds to the i'th atom of the
 second molecule. You can also select atoms in either the <em>slave</em>
 or the <em>master</em> molecule. The <em>slave</em> is the molecule to be
 aligned. If the molecules have different numbers of atoms, the only
 way to align molecules is to select atoms in both molecules.</p>
</div>
<p></p>
<p></p><div class="hxconnections">
<a name="A2"></a><h3>Connections</h3>

<div class="hxconnection_optional">
 <span class="headline">Data</span> <span class="connectiontype">[optional]</span> 

<div class="body"><p>The molecule may be connected to a <i>molecular dynamics
 trajectory</i>. If so, a <em>Time</em> port will appear in the
 <a class="link" href="../usersguide/mainwin.html#A4">Properties Area</a> which enables you to load new time
 steps into the molecule.</p>
<p> <a name="A3"></a>
</p></div></div>
<div class="hxconnection_optional">
 <span class="headline">Time</span> <span class="connectiontype">[optional]</span> 

<div class="body"><p>Sometimes it is necessary to synchronize two or more time-dependent
 data objects, such as a molecule from a trajectory and the
 corresponding electrostatic field. In this case, the <em>Time</em> port
 can be connected to a <em>Time</em> object.</p>
<p> <a name="A4"></a>
</p></div></div>
<a name="A5"></a>
<div class="hxconnection_optional">
 <span class="headline">AlignMaster</span> <span class="connectiontype">[optional]</span> 

<div class="body"></div></div>
<div class="hxconnection_optional">
 <span class="headline">PrecomputedAlignment</span> <span class="connectiontype">[optional]</span> 

<div class="body"><p>You find the description of the two ports above are described in the
 section on <a class="link" href="molpositioning.html#A1">alignment of molecules</a>.
</p></div></div>
</div>
<p></p>
<p></p><div class="hxports">
<a name="A6"></a><h3>Ports</h3>

<div class="hxport">
 <span class="headline">Time</span>  
<div class="portsnapshot"><img src="HxMolecule_time.png" /><br /></div>
<div class="body"><p>The <i>Time</i> port is visible only if the molecule object is time
 dependent, i.e., it is connected to a molecular dynamics trajectory.
 In the slider you can select a specific time step. If you wish to
 step through the trajectory, use the buttons next to the slider. For
 continuous animation, use the outer buttons.</p>
<p> <a name="A7"></a>
</p></div></div>
<div class="hxport">
 <span class="headline">Alignment</span>  
<div class="portsnapshot"><img src="HxMolecule_transformation.png" /><br /><img src="HxMolecule_selection.png" /><br /><img src="HxMolecule_selectionInfo.png" /><br /></div>
<div class="body"><p>You will find the description of the above three ports in the section
 on <a class="link" href="molpositioning.html#A1">alignment of molecules</a>.</p>
<p> <a name="A8"></a>
</p></div></div>
<div class="hxport">
 <span class="headline">Selection Browser</span>  
<div class="portsnapshot"><img src="HxMolecule_selectionBrowser.png" /><br /></div>
<div class="body"><p>Press this button if you wish to open the selection browser for this
 molecule.</p>
<p> <a name="A9"></a>
</p></div></div>
<div class="hxport">
 <span class="headline">Transform</span>  
<div class="portsnapshot"><img src="HxMolecule_transform.png" /><br /></div>
<div class="body"><p>If you want to transform all atom coordinates of the molecule by
 the current transformation in order to save the molecule with the
 transformed coordinates to a file, press the <em>Apply</em> button.
 As long as the <em>Transform Editor</em> is active you can also undo
 your actions.</p>
</div></div>
</div>
<p>
</p><div class="hxcommands">
<a name="A10"></a><h3>Commands</h3>
<p>
 
You can use the <a class="link" href="../usersguide/console.html#A1">console window</a> for entering
 additional commands for molecules.
 The easiest way to use the console for a molecule is to click on the
 molecule object in the Pool and then press the <tt>TAB</tt> key
 in the console window to display the name of the selected
 object. After the name you can use the following selection
 commands:</p>

<div class="hxcommand">
  <span class="headline">list</span>  

<div class="body"><p>
 Lists all levels defined for the molecule.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">list &lt;levelName&gt;</span>  

<div class="body"><p>
 Lists all groups of level <em>levelName</em>.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">list &lt;levelName/[groupName|groupRange]&gt;</span>  

<div class="body"><p> Prints information about a single
 group or multiple groups specified by a group range.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">define &lt;levelName/groupName&gt; &lt;groups&gt;</span>  

<div class="body"><p>
 This command defines a new group <em>groupName</em> in the level
 <em>levelName</em>. If <em>levelName</em> does not yet exist, it
 will be added as a new level. <em>groups</em> is a list of groups
 of possibly different levels. Groups can be specified using
 ranges.<br />

 Example: The following command defines group <em>one</em> in
 level <em>test</em> consisting of atoms 1 and 3 and residues 4 to
 6.
 </p>
<ul><li><pre> define test/one atoms/1 atoms/3 residues/4-6
</pre></li>
</ul>
</div></div>
<div class="hxcommand">
  <span class="headline">defregexp &lt;levelName/groupName&gt; &lt;atomExpr&gt;</span>  

<div class="body"><p>
 Define a new group by using
 <a class="link" href="molatomexpressions.html#A1">atom expressions</a>.
 The new group will only contain atoms, i.e., no higher level groups.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">applyTransform</span>  

<div class="body"><p>
 The current transform is applied to the molecule, i.e., the
 original atomic positions are replaced by the transformed
 positions.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">generateSmiles</span>  

<div class="body"><p>
Returns SMILES string of molecule.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">alignBySmartsMatching</span>  

<div class="body"><p>
Aligns molecule to &lt;mol2&gt; by first matching the given smarts string to both molecules
and then using this matching for computing the transformation yielding the lowest rmsd
between the matched groups. If either of the molecules did not have a matching
nothing will be done and 0 is returned. Else 1 is returned.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">computeRMSD</span>  

<div class="body"><p>
Returns the root mean square distance between the molecule and another molecule.
The molecules must have identical topologies (this means the sequence of the atoms must be
identical too).</p>
</div></div>
<div class="hxcommand">
  <span class="headline">savePQR</span>  

<div class="body"><p>
Saves the molcule in PQR format (pdb format with added partial charges and radii).
If no attribute names are supplied as parameters, the default values 'charge' and 'radius' will be used.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">getMolWeight</span>  

<div class="body"><p>
Returns the molecular weight of the molecule. The computation will consider implicit hydrogens in
the atoms/implicit_hnum attribute.
</p></div></div><p>
<b>Viewer Commands:</b></p>
<p>The following commands allow to connect different kinds of viewer modules to the molecule from
the command line:</p>

<div class="hxcommand">
  <span class="headline">showSurface</span>  

<div class="body"><p>
Creates CompMolSurface and MolSurfaceView modules to show the molecular surface.
No return value.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">showSticks</span>  

<div class="body"><p>
Creates a MolecuelView module to show the molecule in sticks representation.
No return value.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">showHBonds</span>  

<div class="body"><p>
Creates a HBondsView module to show all hydrogens bonds.
No return value.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">showHBondsTo</span>  

<div class="body"><p>
Create a new molecule in the object pool containing all hyrdogens bonds between
&lt;molecule&gt; and &lt;molecule2&gt;. A HBondView will be connected to
this molecule to show the hydrogen bonds. Returns name of new molecule.
</p></div></div><p>
<b>Selection Commands:</b></p>

<div class="hxcommand">
  <span class="headline">showBroswer</span>  

<div class="body"><p>
Shows the selection browser.
No return value.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">sel &lt;atomExpr&gt;</span>  

<div class="body"><p>
 Lets you add atoms specified by <em>atomExpr</em> to the current
 selection.<br />

 Example: The following command will select all carbon atoms
 which are located in helices.
 </p>
<ul><li><pre> sel a=C AND s/type=helix </pre></li>
</ul><p>
 There is also a <a class="link" href="molglobaltclcommands.html#A1">global command</a> 'sel' which applies the atom expressions to all
 molecules in the object pool.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">desel &lt;atomExpr&gt;</span>  

<div class="body"><p>
 Lets you remove atoms from the selection.
 Like for 'sel' there is also a global 'desel' command operating on all molecules.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">selAtom &lt;atomIx&gt;</span>  

<div class="body"><p>
 Selects atom with given index.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">deselAtom &lt;atomIx&gt;</span>  

<div class="body"><p>
 Deselects atom with given index.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">isAtomSelected &lt;atomIx&gt;</span>  

<div class="body"><p>
 Returns 1 if atom with given index is selected, 0 otherwise.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">getNumSelection</span>  

<div class="body"><p>
 Returns number of selected atoms.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">invertSelection</span>  

<div class="body"><p>
 Atoms which are selected will become unselected and vice versa.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">selWithinSelection &lt;objectname&gt; &lt;range&gt;</span>  

<div class="body"><p>
 Atom expressions only work on single molecule objects. This command allows the WITHIN operator
 of atom expressions to be used in relation to another molecule object in the object pool. The command will select
 all atoms that are within &lt;range&gt; Angstroem from any atom selected in the molecule &lt;objectname&gt;.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">selWithinCoordinate &lt;x&gt; &lt;y&gt; &lt;z&gt; &lt;range&gt;</span>  

<div class="body"><p>
 Will select all atoms which are within &lt;range&gt; Angstroem from the given coordinate.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">expandSelectionToGroupsOfLevel &lt;levelIx&gt;</span>  

<div class="body"><p>
 Will expand selection to all groups of given level for which at least a single atom is selected.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">selFromFilter [&lt;objectName&gt;]</span>  

<div class="body"><p>
 Selects all atoms which are not hidden in the filter of the given downstream object.
 If the name of the object is omitted it will select all atoms which are not hidden
 in any of the downstream object filters. No return value.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">selToFilter [&lt;objectName&gt;]</span>  

<div class="body"><p>
 Sets filter of the given object to the currently selected atoms.
 If the name of the object is omitted it will apply to all downstream objects.
 No return value.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">printSelection</span>  

<div class="body"><p>
Prints a lists of all atoms which are selected.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">printSelectionByAtoms</span>  

<div class="body"><p>
Same as printSelection.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">printSelectionByResidues</span>  

<div class="body"><p>
Prints a lists of all residues which are entirely or partially selected.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">printSelectionByLevel &lt;levelName&gt;</span>  

<div class="body"><p>
Prints a lists of all groups of the given level which are entirely or partially selected.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">getSelectionCenter</span>  

<div class="body"><p>
Computes center of gravity of selected atoms and returns x,y, and z coordinate.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">getSelectionBBox</span>  

<div class="body"><p>
Computes the bounding box surrounding all selected atoms and returns it as xmin,xmax,ymin,ymax,zmin,zmax.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">addSet &lt;name&gt;</span>  

<div class="body"><p>
Adds current selection as a new group to the 'sets' level. The name attribute of the new group will have
the value &lt;name&gt;. This set can be used to return to the current selection at a later time.
Note that there is also a <a class="link" href="molglobaltclcommands.html#A1">global command</a> which will apply addSet to all molecules in the object pool.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">removeSet &lt;name&gt;</span>  

<div class="body"><p>
Remove set of name &lt;name&gt; from 'sets' level.
No return value.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">useSet &lt;name&gt;</span>  

<div class="body"><p>
Resets selection to atoms contained in the group in 'sets' level which has the name attribute value &lt;name&gt;.
No return value. Like for addSet, there is also a global useSet command.
</p></div></div><p>
<b>Labeling commands:</b></p>
<p>The following command allow to modify labels of groups. They work on all downstream MolLabel modules.</p>

<div class="hxcommand">
  <span class="headline">setGroupLabel &lt;levelIx&gt; &lt;groupIx&gt; &lt;label&gt;</span>  

<div class="body"><p>
Attaches a label to a group. This is done by setting the label in a create MolLabel module. If no module exits,
a new MolLabel will be created. No return value.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">setGroupLabelSize [&lt;levelIx&gt;] &lt;size&gt;</span>  

<div class="body"><p>
This command works on all attached MolLabel modules and sets the font size of all labels for the given level.
If the level index is omitted, it will be applied to all levels. No return value.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">setGroupLabelColor [&lt;levelIx&gt;] &lt;red|green|blue|yellow|white|black&gt;</span>  

<div class="body"><p>
This command works on all attached MolLabel modules and sets the color of all labels for the given level.
If the level index is omitted, it will be applied to all levels. No return value.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">setGroupLabelColorRGB [&lt;levelIx&gt;] &lt;r&gt; &lt;g&gt; &lt;b&gt;</span>  

<div class="body"><p>
Same as setGroupLabelColor but instead of using a text string defining a color, the red green blue
value of the color (each within the range [0,1]) is specified. No return value.
</p></div></div><p>
<b>Editing levels:</b>
The following commands allow to modify the level hierarchy of the molecule.
The commands work index based. Removing levels may
invalidate locally stored level indices.</p>

<div class="hxcommand">
  <span class="headline">getLevelIx &lt;levelName&gt;</span>  

<div class="body"><p>
Returns index of level with name &lt;levelName&gt;, 0 if no such level.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">getLevelName &lt;levelIx&gt;</span>  

<div class="body"><p>
Returns name of level with index &lt;levelIx&gt;.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">getNumLevels</span>  

<div class="body"><p>
Returns number of levels.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">addLevel &lt;levelName&gt; &lt;referenceLevelIx&gt;</span>  

<div class="body"><p>
Adds level of name &lt;levelName&gt; which references the level with index &lt;referenceLevelIx&gt;.
Returns index of new level</p>
</div></div>
<div class="hxcommand">
  <span class="headline">removeLevel &lt;levelIx&gt;</span>  

<div class="body"><p>
Removes level with index levelIx. All dependant levels will also be removed. Level indices might change after using this command.
No return value.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">getReferenceLevelIx &lt;levelIx&gt;</span>  

<div class="body"><p>
Returns the index of the referenced level of level &lt;levelIx&gt;. Returns 0 if it does not reference another level.
</p></div></div><p>
<b>Editing groups of a level:</b>
The following commands allow to modify the groups that are part of a level.
The commands work index based. Removing groups may
invalidate locally stored group indices.</p>

<div class="hxcommand">
  <span class="headline">getNumGroups &lt;levelIx&gt;</span>  

<div class="body"><p>
Returns number of groups of level with index &lt;levelIx&gt;.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">getNumGroupElements &lt;levelIx&gt; &lt;groupIx&gt;</span>  

<div class="body"><p>
Returns number of elements of group &lt;groupIx&gt; of level &lt;levelIx&gt;.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">getGroupElement &lt;levelIx&gt; &lt;groupIx&gt; &lt;elementIx&gt;</span>  

<div class="body"><p>
Returns index of &lt;elementIx&gt;'th group which is referenced by group &lt;groupIx&gt; of level &lt;levelIx&gt;.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">getGroupElements &lt;levelIx&gt; &lt;groupIx&gt;</span>  

<div class="body"><p>
Returns list which contains indices of groups which are referenced by group &lt;groupIx&gt; of level &lt;levelIx&gt;.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">removeGroupElement &lt;levelIx&gt; &lt;groupIx&gt; &lt;elementIx&gt;</span>  

<div class="body"><p>
Removes &lt;elementIx&gt;'th reference of group &lt;groupIx&gt; of level &lt;levelIx&gt;. Element indices of the group will change after using this command.
No return value.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">setGroupElement &lt;levelIx&gt; &lt;groupIx&gt; &lt;elementIx&gt; &lt;refGroupIx&gt;</span>  

<div class="body"><p>
Sets the reference of the &lt;elementIx&gt;'th element of the group &lt;groupIx&gt; of level &lt;levelIx&gt; to the group &lt;refGroupIx&gt;.
No return value.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">addGroupElement &lt;levelIx&gt; &lt;groupIx&gt; &lt;refGroupIx&gt;</span>  

<div class="body"><p>
Add the reference &lt;refGroupIx&gt; at the end of the of the elements of the group &lt;groupIx&gt; of level &lt;levelIx&gt;.
No return value.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">removeGroup &lt;levelIx&gt; &lt;groupIx&gt;</span>  

<div class="body"><p>
Remove the group &lt;groupIx&gt; from index &lt;levelIx&gt;. Will also remove all dependant groups and attribute entries.
No return value.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">removeGroups &lt;levelIx&gt; &lt;groupIx1&gt; &lt;groupIx2&gt; ...</span>  

<div class="body"><p>
Remove the given groups from level &lt;levelIx&gt;. Will also remove all dependant groups and attribute entries.
No return value.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">restrictToGroups &lt;levelIx&gt; &lt;groupIx1&gt; &lt;groupIx2&gt; ...</span>  

<div class="body"><p>
Remove all groups of level &lt;levelIx&gt; except for the given groups.
Will also remove all dependant groups and attribute entries.
No return value.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">removeSelection</span>  

<div class="body"><p>
Remove all atoms which are selected and all dependant groups.
No return value.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">restrictToSelection</span>  

<div class="body"><p>
Remove all atoms which are not selected and all dependant groups.
No return value.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">copySelection</span>  

<div class="body"><p>
Copies the selected atoms into a new Molecule object. The label of the new object in the object pool will be returned.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">splitSelection</span>  

<div class="body"><p>
Copies the selected atoms into a new Molecule object and removed them from the current.
The label of the new object in the object pool will be returned.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">addMolecule &lt;objectname&gt;</span>  

<div class="body"><p>
Adds the molecule object in the object pool with the given name.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">addMoleculeSelection &lt;objectname&gt;</span>  

<div class="body"><p>
Adds all selected groups of the molecule object in the object pool with the given name.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">addGroup &lt;levelIx&gt;</span>  

<div class="body"><p>
Adds group to level &lt;levelIx&gt;.
Returns index of new group.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">addGroupSelection &lt;levelIx&gt;</span>  

<div class="body"><p>
Adds group to level &lt;levelIx&gt; and sets its group elements to the currently selected atoms.
Returns index of new group.
</p></div></div><p>
<b>Editing attributes:</b>
The following commands allow to modify attributes of groups.
The commands work index based. Removing attributes may
invalidate locally stored attribute indices.</p>

<div class="hxcommand">
  <span class="headline">getNumAttributes &lt;levelIx&gt; </span>  

<div class="body"><p>
Return number of attributes of level &lt;levelIx&gt;.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">getAttributeIx &lt;levelIx&gt; &lt;attrName&gt;</span>  

<div class="body"><p>
Return index of attribute &lt;attrName&gt; of level &lt;levelIx&gt;.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">getAttributeName &lt;levelIx&gt; &lt;attrIx&gt;</span>  

<div class="body"><p>
Return name of attribute &lt;attrIx&gt; of level &lt;levelIx&gt;.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">addStringAttribute &lt;levelIx&gt; &lt;attrName&gt;</span>  

<div class="body"><p>
Adds a string attribute of name &lt;attrName&gt; to level &lt;levelIx&gt;.
Returns the index of new attribute. Returns 0 if not successful.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">addIntegerAttribute &lt;levelIx&gt; &lt;attrName&gt;</span>  

<div class="body"><p>
Adds an integer attribute of name &lt;attrName&gt; to level &lt;levelIx&gt;.
Returns the index of new attribute. Returns 0 if not successful.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">addFloatAttribute &lt;levelIx&gt; &lt;attrName&gt;</span>  

<div class="body"><p>
Adds a float attribute of name &lt;attrName&gt; to level &lt;levelIx&gt;.
Returns the index of new attribute. Returns 0 if not successful.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">removeAttribute &lt;levelIx&gt; &lt;attrIx&gt;</span>  

<div class="body"><p>
Remove attribute &lt;attrIx&gt; from level &lt;levelIx&gt;.
No return value.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">setAttributeName &lt;levelIx&gt; &lt;attrIx&gt; &lt;attrName&gt;</span>  

<div class="body"><p>
Sets name of attribute &lt;attrIx&gt; of level &lt;levelIx&gt; to &lt;attrName&gt;.
No return value.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">getAttributeValue &lt;levelIx&gt; &lt;attrIx&gt; &lt;groupIx&gt;</span>  

<div class="body"><p>
Returns value of attribute &lt;attrIx&gt; of level &lt;levelIx&gt; for the group &lt;groupIx&gt;.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">setAttributeValue &lt;levelIx&gt; &lt;attrIx&gt; &lt;groupIx&gt; &lt;value&gt;</span>  

<div class="body"><p>
Sets value of attribute &lt;attrIx&gt; of level &lt;levelIx&gt; for the group &lt;groupIx&gt; to &lt;value&gt;.
</p></div></div><p>
<b>Editing coordinates:</b>
</p>
<div class="hxcommand">
  <span class="headline">getCoordinate &lt;atomIx&gt;</span>  

<div class="body"><p>
Returns the coordinate of atom &lt;atomIx&gt;.
No return value.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">setCoordinate &lt;atomIx&gt; &lt;x&gt; &lt;y&gt; &lt;z&gt;</span>  

<div class="body"><p>
Sets the coordinate of atom &lt;atomIx&gt;.
No return value.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">perturbCoordinates [&lt;max&gt;]</span>  

<div class="body"><p>
Applies a random perturbation to each coordinate.
The perturbation will be uniformly distributed in the range [-max,max].
If the parameter is omitted 0.001 will be used.
No return value.
</p></div></div><p>
<b>Editing data entries:</b>
The following commands allow to modify data entries of the molecule.
The commands work either by using the data entry's index or its name.
Removing entries may invalidate locally stored data indices.</p>

<div class="hxcommand">
  <span class="headline">setData &lt;dataID&gt; &lt;value&gt;</span>  

<div class="body"><p>
Adds a data entry with the given name and value. The name
must contain at least one letter.
If the entry already exists &lt;dataID&gt; can be either its name or its index
and the value will be reset to the new value.
No return value.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">removeData &lt;dataID&gt;</span>  

<div class="body"><p>
Removes a data entry. The &lt;dataId&gt; may either be the
entry's name or its index.
No return value.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">getDataValue &lt;dataID&gt;</span>  

<div class="body"><p>
Returns the value of a data entry. The &lt;dataId&gt; may either be the
entry's name or its index.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">getDataName &lt;dataIx&gt;</span>  

<div class="body"><p>
Returns the name of a data entry with the given index.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">getDataIx &lt;dataName&gt;</span>  

<div class="body"><p>
Returns the index of a data entry with the given name.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">getNumData</span>  

<div class="body"><p>
Returns the number of data entries.
</p></div></div><p>
<b>Miscellaneous editing commands:</b>
</p>
<div class="hxcommand">
  <span class="headline">addHydrogens</span>  

<div class="body"><p>
Add hydrogens until the valences of all atoms are saturated.
No return value.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">removeHydrogens</span>  

<div class="body"><p>
Removes hydrogens.
No return value.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">removeNonPolarHydrogens</span>  

<div class="body"><p>
Removes hydrogens on non polar heavy atoms.
No return value.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">splitByLevel &lt;levelIx&gt;</span>  

<div class="body"><p>
Splits molecule by putting each group of the given level into a seperate MolTrajectory.
Returns the name of the MolTrajectoryBundle object which is created.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">splitByAttribute &lt;levelIx&gt; &lt;attributeIx&gt;</span>  

<div class="body"><p>
Splits molecule by putting all groups of the given level that have the same attribute
value into a seperate MolTrajectory. If the attribute is the index attribute of the level
this command does the same as the splitByLevel command.
Returns the name of the MolTrajectoryBundle object which is created.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">cleanBonds</span>  

<div class="body"><p>
Some imported molecule structures have duplicate entries for the same bond.
This leads to problems in some algorithms. This commands removes such duplicates.
No return value.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">addMMFFParameterization</span>  

<div class="body"><p>
Adds parameters of MMFF94 force field.
No return value.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">addMassAttribute [&lt;name&gt;]</span>  

<div class="body"><p>
Adds an atom attribute containing the atomic mass.
Specifying the name of the new attribute is optional. If no name is
given 'mass' will be used.
No return value.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">addRadiusAttribute [&lt;name&gt;]</span>  

<div class="body"><p>
Adds an atom attribute containing the standard van der Waals radius.
Specifying the name of the new attribute is optional. If no name is
given 'radius' will be used.
No return value.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">removeWater</span>  

<div class="body"><p>
Removes all water molecules and unbonded oxygens and hydrogens.
No return value.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">assignKekuleBondOrders</span>  

<div class="body"><p>
Converts all aromatic bond orders to single and double to create a kekule structure.
No return value.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">assignNonKekuleBondOrders</span>  

<div class="body"><p>
Reverses the kekule structure assignment.
No return value.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">computeHBonds [&lt;maxDistDA&gt; &lt;minAngleDAX&gt;]</span>  

<div class="body"><p>
Will create a level 'hBonds' containing potential hydrogen bonds.
The optional parameters &lt;maxDistDA&gt; determines the maximum distance
between donor and acceptor atom and &lt;minAngleDAX&gt; the minimum angle
between donor, acceptor, and a heavy atom bonded to the acceptor.
The default values are 3.9 and 120.
For more fine grained control and more details look at the documentation of
the <a class="link" href="../hxmolstructure/HxComputeHBonds.html#A1">ComputeHBonds</a> module.
No return value.
</p></div></div><p>
<b>Protein specific editing commands:</b></p>

<div class="hxcommand">
  <span class="headline">alignToProtein &lt;mol2&gt;</span>  

<div class="body"><p>
Aligns this molecule to another molecule in the object pool with the given label.
The alignment will be accomplished by residue sequence alignment with the CompSeqAlign module.
Both molecules need to contain the residues/type attribute with 3 letter code amino acid types.
No return value.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">alignProtein &lt;mol2&gt;</span>  

<div class="body"><p>
Aligns a molecule in the object pool with the given label to this molecule.
Both molecules need to contain the residues/type attribute with 3 letter code amino acid types.
No return value.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">applyBiomt</span>  

<div class="body"><p>
PDB files may contain information how to generate multimers from a single unit
in BIOMT REMARK records. These records are read in as a PDB_REMARK data field.
The applyBiomt command will parse all BIOMT matrices in this field, duplicate the 
structure, and apply the specified matrix transformations.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">appendAminoAcid &lt;index&gt; &lt;terminus&gt; &lt;code&gt;</span>  

<div class="body"><p>
Append a new amino acid to the amino acid that is the index'th group in the residues level. The type of the new amino acid must be given in the 3 letter code.
Terminus must be 'C' or 'N' depending on which terminus the old index'th residue the new residue is appended.
For this command to work, the atoms/type attribute must contain the pdb atom types and the
residues/type and bonds/type attribute must exist.
No return value.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">substituteAminoAcid &lt;index&gt; &lt;code&gt;</span>  

<div class="body"><p>
Substitutes the amino acid that is the index'th group in the residues level with the amino acid of the given
3-letter code. For this command to work, the atoms/type attribute must contain the pdb atom types and the
residues/type and bonds/type attribute must exist.
No return value.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">mutatePDB &lt;mutationString&gt;</span>  

<div class="body"><p>
This command acts similar as the substituteAminoAcid command except that the syntax of the substitution definition
differs to make this command more easily usable with pdb proteins. The mutation string is a concatenation
of the original type, the pdb index and the new type (example LYS34ASP).
Both 3 or 1-letter code may be used (L34H). The index refers to the pdb_index attribute.
If the original type is left out, the amino acid with the given index will be substituted
regardless of type (example 34H). If the index is left out, all amino
acids of the given original type will be substituted with the new type.
In this latter case, 3-letter codes need to be used (example LYSASP).
A set of substitutions can be applied at once by seperating them by whitespaces and
enclosing the set in curly brackets
(example &#123;L23H G24S&#125;).
No return value.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">getAminoAcidChirality [&lt;index&gt;]</span>  

<div class="body"><p>
Returns whether amino acid &lt;index&gt; is an 'L' or 'D' isomer. Returns 'N' if residue is not an
amino acid or if required attributes are missing. If &lt;index&gt; is omitted a string
containing 'L', 'D' or 'N' for each residue will be returned.
For this command to work, the atoms/type attribute must contain the pdb atom types
and the residues level must exist containing the amino acids.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">setAminoAcidChirality &lt;index&gt; &lt;chiralCode&gt;</span>  

<div class="body"><p>
Sets amino acid to 'L' or 'D' isomer.
For this command to work, the atoms/type attribute must contain the pdb atom types
and the residues level must exist containing the amino acids. No return value.</p>
</div></div>
<div class="hxcommand">
  <span class="headline">getProteinFASTASequence</span>  

<div class="body"><p>
Returns the protein sequence in FASTA format for all convertible chains.
The topology must have chains and residues levels and contain the attributes 
chains/name, residues/type, and residues/name. 
All residues whose name starts with 'HET' will be cut out.
If a chain contains any other residues whose type cannot 
be converted from 3 letter code to 1 letter code, it will be skipped.</p>
</div></div>
</div>
<p></p>
<p></p>
<p>

</p></body>
</html>
